pkg/core/tekton/builder/meta.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2019 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package builder
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// MetaOp is an interface that is used in other builders.
+// Other builders should have a Meta function that accepts ...MetaOp where ObjectMetaOp/TypeMetaOp are the underlying type.
+type MetaOp interface{}
+
+// ObjectMetaOp is an operation which modifies the ObjectMeta.
+type ObjectMetaOp func(m *metav1.ObjectMeta)
+
+// TypeMetaOp is an operation which modifies the TypeMeta.
+type TypeMetaOp func(m *metav1.TypeMeta)
+
+// Label adds a single label to the ObjectMeta.
+func Label(key, value string) ObjectMetaOp {
```

Comment:
Indeed in you example the code is clearly more readable. However I am note so sure about the code in here as we would have, for example:
`prb.PipelineRunMeta(builder.WithLabel("foor", "bar"), builder.WithAnnotation("alpha", "beta"))`

instead of:
`prb.PipelineRunMeta(builder.Label("foor", "bar"), builder.Annotation("alpha", "beta"))`

IMO the latter looks better as the first one gives the idea that the builder will create "something" with the given label, and another "thing" with the given annotation. While the last one gives the idea that the builder is building a `Label` and `Annotation` object which is correct.

Reply:
The last one gives the impression that those are objects (label and annotation) rather than operations, but I guess it's just a matter of interpretation, so I'm fine with it this way if you are.